File: app/src/main/java/com/tomclaw/drawa/tools/Marker.java
package com.tomclaw.drawa.tools;

import android.graphics.DashPathEffect;
import android.graphics.Paint;
import android.graphics.Path;

import org.androidannotations.annotations.EBean;

import java.util.Random;

/**
 * Created by solkin on 17.03.17.
 */
@EBean
public class Marker extends Tool {

    private static final int DOT_RADIUS = 4;

    private int startX, startY;
    private int prevX, prevY;
    private Path path;
    private Random random;

    Marker() {
    }

    @Override
    public void onInitialize() {
        this.path = new Path();
        this.random = new Random(System.currentTimeMillis());
    }

    @Override
    Paint initPaint() {
        Paint paint = new Paint(Paint.ANTI_ALIAS_FLAG);
        paint.setStyle(Paint.Style.STROKE);
        paint.setStrokeJoin(Paint.Join.MITER);
        paint.setStrokeCap(Paint.Cap.BUTT);
        paint.setPathEffect(new DashPathEffect(new float[]{2, 0}, 0));
        return paint;
    }

    @Override
    int getAlpha() {
        return 0x50;
    }

    @Override
    public void onTouchDown(int x, int y) {
        startX = x;
        startY = y;
        path.moveTo(x, y);
        path.lineTo(x, y);

        prevX = x;
        prevY = y;

        drawPath(path);
    }

    @Override
    public void onTouchMove(int x, int y) {
        if (path.isEmpty()) {
            path.moveTo(prevX, prevY);
        }
        path.lineTo(x, y);

        prevX = x;
        prevY = y;

        drawPath(path);
    }

    @Override
    public void onTouchUp(int x, int y) {
        if (path.isEmpty()) {
            path.moveTo(prevX, prevY);
        }
        if (x == startX && y == startY) {
            for (int c = 0; c < 3; c++) {
                path.lineTo(randomizeCoordinate(x), randomizeCoordinate(y));
                drawPath(path);
            }
        } else {
            path.lineTo(x, y);
        }

        drawPath(path);

        prevX = 0;
        prevY = 0;
    }

    @Override
    public void onDraw() {
        path.reset();
    }

    @Override
    public byte getType() {
        return TYPE_MARKER;
    }

    private int randomizeCoordinate(int value) {
        return value + random.nextInt(DOT_RADIUS + 1) - DOT_RADIUS / 2;
    }
}

File: app/src/main/java/com/tomclaw/drawa/draw/tools/Marker.kt
package com.tomclaw.drawa.draw.tools

import android.graphics.DashPathEffect
import android.graphics.Paint
import android.graphics.Path
import java.util.Random

class Marker : Tool() {

    private var startX: Int = 0
    private var startY: Int = 0
    private var prevX: Int = 0
    private var prevY: Int = 0
    private var path = Path()
    private var random = Random()

    override val alpha = 0x50
    override val type = TYPE_MARKER

    override fun initPaint() = Paint().apply {
        isAntiAlias = true
        isDither = true
        style = Paint.Style.STROKE
        strokeJoin = Paint.Join.MITER
        strokeCap = Paint.Cap.BUTT
        pathEffect = DashPathEffect(floatArrayOf(2f, 0f), 0f)
    }

    override fun onTouchDown(x: Int, y: Int) {
        resetRadius()

        startX = x
        startY = y

        path.moveTo(x.toFloat(), y.toFloat())
        path.lineTo(x.toFloat(), y.toFloat())

        prevX = x
        prevY = y

        drawPath(path)
    }

    override fun onTouchMove(x: Int, y: Int) {
        if (path.isEmpty) {
            path.moveTo(prevX.toFloat(), prevY.toFloat())
        }
        path.lineTo(x.toFloat(), y.toFloat())

        prevX = x
        prevY = y

        drawPath(path)
    }

    override fun onTouchUp(x: Int, y: Int) {
        if (path.isEmpty) {
            path.moveTo(prevX.toFloat(), prevY.toFloat())
        }
        if (x == startX && y == startY) {
            for (c in 0..2) {
                path.lineTo(randomizeCoordinate(x).toFloat(), randomizeCoordinate(y).toFloat())
                drawPath(path)
            }
        } else {
            path.lineTo(x.toFloat(), y.toFloat())
        }

        drawPath(path)

        prevX = 0
        prevY = 0
    }

    override fun onDraw() = path.reset()

    private fun randomizeCoordinate(value: Int) =
            value + random.nextInt(DOT_RADIUS + 1) - DOT_RADIUS / 2

}

private const val DOT_RADIUS = 4
